### DISCUSSION TITLE
Plotting projection from E3SM tutorial: NotImplementedError

### DISCUSSION THREAD

**Discussion by pgoddard08** [UXARRAY/uxarray#920] [Q&A]
Hello,

I have a 1D unstructured uxdataarray that I am trying to plot using instructions from https://github.com/xCDAT/xcdat-validation/blob/b7c6dba828b39dc72e1630a671b1f0f4e67c33d8/demos/2024-e3sm-workshop/uxarray_practicum_notebook.ipynb.

Here is my code to plot Z500 for a specific date:
```
date = '2018-02-01'

xxx = e3sm_z500_12GMT.sel(time=e3sm_z500_12GMT.time.dt.strftime('%Y-%m-%d') == date).squeeze()

projection = projection = ccrs.Robinson()

xxx.plot.rasterize(
    method="point",
    projection=projection,
    backend="matplotlib",
    pixel_ratio=8.0,
    fig_size=400,
) * features(projection)
```

This results in the error:
```
NotImplementedError: An error occurred while calling the from_dict method registered to the pandas backend.
Original Message: from_dict doesn't currently support Dask collections as inputs. Objects of type {<class 'dask.array.core.Array'>} were given in the input dict.
```

Thanks for the help!
Paul

**Comment by pgoddard08** [comment]
Here is the print out for the variable:
![image](https://github.com/user-attachments/assets/1edb44cd-0883-4d2e-b161-b9891166eb3e)

**Reply by philipc2** [reply]
Hi @pgoddard08 

Can you add the following line right before you declare a projection? This should work as a temporary fix. 

```Python
xxx = xxx.compute()
```

It appears that there's an issue with the data being represented as a Dask Array for the Point Visualization. I'll try to recreate this, since this is most likely not intended behavior.

**Reply by philipc2** [reply]
I'd like to share our Matplotlib visualization support as well, as it supports Polygon plots with projections. 

https://uxarray.readthedocs.io/en/latest/user-guide/mpl.html#geographic-projections-elements


Adapting the code you provided:

```Python

mport uxarray as ux
import matplotlib.pyplot as plt
import cartopy.crs as ccrs
import cartopy.feature as cfeature

date = '2018-02-01'

xxx = e3sm_z500_12GMT.sel(time=e3sm_z500_12GMT.time.dt.strftime('%Y-%m-%d') == date).squeeze().compute()

projection = ccrs.Robinson()

pc = xxx..to_polycollection(projection=projection, override=True)

pc.set_antialiased(False)
pc.set_cmap("plasma")

fig, ax = plt.subplots(
    1,
    1,
    figsize=(10, 5),
    facecolor="w",
    constrained_layout=True,
    subplot_kw=dict(projection=projection),
)

# add geographic features
ax.add_feature(cfeature.COASTLINE)
ax.add_feature(cfeature.BORDERS)

ax.add_collection(pc)
ax.set_global()
plt.title("Plot with Matplotlib")
```

**Reply by pgoddard08** [reply]
Great - that got me past the error. The output was not what I expected, but this is my first day using UxArray so I'll play around some more.
Output:
![image](https://github.com/user-attachments/assets/34a311a9-bf9f-4b46-9325-bfd97ec6fe33)

**Reply by philipc2** [reply]
> Great - that got me past the error. The output was not what I expected, but this is my first day using UxArray so I'll play around some more. Output: ![image](https://private-user-images.githubusercontent.com/73358072/362855082-34a311a9-bf9f-4b46-9325-bfd97ec6fe33.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ5NTg0NjAsIm5iZiI6MTcyNDk1ODE2MCwicGF0aCI6Ii83MzM1ODA3Mi8zNjI4NTUwODItMzRhMzExYTktYmY5Zi00YjQ2LTkzMjUtYmZkOTdlYzZmZTMzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI5VDE5MDI0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU3NmExNTZkNzAyYTQ2ZTY4OGU5ZTZiNDU4MTA4ZjVkOGM2Yzk0NzllZTIzMWFmZjUzNDBkYWZlMDE1ZmEyMDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.hSWvUjgWalBIjPpEFaCAfefRCAkRubYHfhYbOQxEG2Y)

That's because of the point raster. The data is being rendered at the center of each face, instead of shading the polygon. I'd suggest:
- Using `method='polygon'` (with no projection)
- Using Matplotlib's PolyCollection (with projections)

This should be the desired results. Let me know if there's anything else I can help with!

**Reply by pgoddard08** [reply]
Hi Philip - please let me know if I should start a new thread with the following questions. Thanks for your help!

Using the directions [here](https://uxarray.readthedocs.io/en/latest/user-guide/mpl.html) as a guide, I am running into a couple errors.

```
date = '2018-02-01'

xxx = e3sm_z500_12GMT.sel(time=e3sm_z500_12GMT.time.dt.strftime('%Y-%m-%d') == date).squeeze().compute()

lc = xxx.uxgrid.to_linecollection()

lc.set_linewidth(0.5)
lc.set_edgecolor("black")

fig, ax = plt.subplots(1, 1, figsize=(10, 5), constrained_layout=True)

ax.set_xlim((-180, 180))
ax.set_ylim((-90, 90))
ax.add_collection(lc)

plt.title("Line Collection Plot")
plt.show()
```

Gives me the error: `RuntimeError: Can not put single artist in more than one figure`


And, `pc = xxx.to_polycollection()`
Gives me the error `TypeError: cannot unpack non-iterable PolyCollection object`

**Reply by philipc2** [reply]
May you share what `uxarray` version you are using? Can run `uxarray.__version__`

**Reply by pgoddard08** [reply]
Thanks, version 2024.04.0 is when I enter `conda list - n myenv` at the command line, but in my jupyter notebook when I run `uxarray.__version__` it says 0.0.0

**Reply by philipc2** [reply]
That might the issue. Can you make sure you're on the latest UXarray release? We've had some major additions to the `PolyCollection` and `LineCollection` methods.

**Reply by pgoddard08** [reply]
Hi, I updated uxarray to version 2024.8.2.  
When I try to plot both lc and pc, I get the error:
`RuntimeError: Can not put single artist in more than one figure`
Thanks for your help!

**Reply by philipc2** [reply]
That issues occurs when you're attempting to use the same collection in more than one figure. You can try setting the `override=True` parameter in the `to_polycollection` and to_linecollection` methods. However, from the code snippet you shared, it doesnt seem like there should be an issue.

**Reply by pgoddard08** [reply]
Ah, ok. Thanks for the information. I tried override=True, but had the same error. I'll look into this further.
